### PR TITLE
Zqs 1096/qsim integration

### PR DESCRIPTION
**Pull request by Boniface316**
## Description

Include description of feature this PR introduces or a bug that it fixes. Include the following information:

Qsim simulator is added to the CirqSimulator.

## Please verify that you have completed the following steps

- [X] I have self-reviewed my code.
- [X] I have included test cases validating introduced feature/fix.
- [X] I have updated documentation.

### CODE REVIEW

File: src/orquestra/integrations/cirq/simulator/_base.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
+import abc
```
Comment (mstechly) [inline src/orquestra/integrations/cirq/simulator/_base.py:1]: This file lacks copyright notice.

File: src/orquestra/integrations/cirq/simulator/qsim_simulator.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    batch_size = sys.maxsize
+
+    def __init__(
+        self,
+        noise_model=None,
+        seed=None,
+        param_resolver=None,
+        qubit_order=cirq.ops.QubitOrder.DEFAULT,
+        circuit_memoization_size: int = 0,
+        qsim_options: Union[None, Dict, qsimcirq.QSimOptions] = None,
```
Comment (mstechly) [inline src/orquestra/integrations/cirq/simulator/qsim_simulator.py:50]: This seems cleaner to me:
```suggestion
        qsim_options: Optional[Union[Dict, qsimcirq.QSimOptions]] = None,
```

File: src/orquestra/integrations/cirq/simulator/qsim_simulator.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        """Return the state vector at the end of the computation
+        Args:
+            circuit (Circuit): the circuit to prepare the state
+            initial_state (StateVector): initial state of the system
+
+        Returns:
+            StateVector: Returns the final state of the computation as ndarray
+
+        Raises:
+            TypeError if this simulator is not QSimSimulator.
```
Comment (mstechly) [inline src/orquestra/integrations/cirq/simulator/qsim_simulator.py:75]: This sounds odd to me. This simulator is `QSimSimulator` by definition, so why would it raise a `TypeError`?

File: src/orquestra/integrations/cirq/simulator/_base.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        return expectation_values_to_real(ExpectationValues(np.asarray(values)))
+
+    @abc.abstractmethod
+    def _get_wavefunction_from_native_circuit(
+        self, circuit: Circuit, initial_state: StateVector
+    ) -> StateVector:
+        pass
+
+    @abc.abstractmethod
+    def _extract_density_matrix(self, result):
```
Comment (mstechly) [inline src/orquestra/integrations/cirq/simulator/_base.py:181]: `result` is pretty vague to me. Should that be `simulation_result` perhaps? Or something else?

File: tests/orquestra/integrations/cirq/simulator/qsimsimulator_test.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    return QSimSimulator()
+
+
+@pytest.fixture()
+def wf_simulator():
+    return QSimSimulator()
+
+
+@pytest.fixture()
+def sampling_simulator():
```
Comment (mstechly) [inline tests/orquestra/integrations/cirq/simulator/qsimsimulator_test.py:30]: Why do we need to define this?

**Review by mstechly — CHANGES_REQUESTED** [review]

File: subtrees/z_quantum_actions/bin/get_next_version.py [hunk 6]
```diff
@@ -7,8 +7,8 @@
 # Reads current project version, bumps "minor", and sets the "next_version" output
 # variable using Github's special stdout syntax.
 
-import sys
 import re
+import sys
```
[the remark this answers is not in the document]
Comment (Boniface316) [reply subtrees/z_quantum_actions/bin/get_next_version.py:11]: It was a change suggested by isort. Should I not let isort handling it inside subtrees?

File: tests/orquestra/integrations/cirq/simulator/qsimsimulator_test.py (hunk 5 above)
Comment (Boniface316) [reply tests/orquestra/integrations/cirq/simulator/qsimsimulator_test.py:30]: I was part of the original code. I can remove this in both cirq and qsim simulator.

File: tests/orquestra/integrations/cirq/simulator/qsimsimulator_test.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    return QSimSimulator()
+
+
+class TestQsimSimulator(QuantumSimulatorTests):
+    def test_setup_basic_simulators(self):
+        simulator = QSimSimulator()
+        assert isinstance(simulator, QSimSimulator)
+        assert simulator.noise_model is None
+
+    def test_run_circuit_and_measure(self):
```
[the remark this answers is not in the document]
Comment (Boniface316) [reply tests/orquestra/integrations/cirq/simulator/qsimsimulator_test.py:40]: QuantumSimulatorTests have variations of this test but not the same.

File: src/orquestra/integrations/cirq/simulator/qsim_simulator.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        simulated_result = self.simulator.simulate(
+            cirq_circuit,
+            param_resolver=self.param_resolver,
+            qubit_order=self.qubit_order,
+            initial_state=initial_state,
+        )
+
+        return simulated_result.final_state_vector
+
+    def _extract_density_matrix(self, result):
```
[the remark this answers is not in the document]
Comment (Boniface316) [reply src/orquestra/integrations/cirq/simulator/qsim_simulator.py:90]: This will be resolved in cirq upgrade

**Review by mstechly — APPROVED** [review]

**Pull request merged**